sklearn/preprocessing/label.py

```diff
@@ -33,6 +34,76 @@
 ]
 
 
+def _nanunique(ar):
+    uniques = np.unique(ar)
+    # np.nan is always sorted last
+    if is_scalar_nan(uniques[-1]):
+        nan_idx = np.searchsorted(uniques, np.nan)
+        return uniques[:nan_idx+1]
+    else:
+        return uniques
+
+
+def _nanencode_numpy(values, uniques=None, encode=False,
+                     missing_value=np.nan):
+    check_values = True
+    if uniques is None:
+        uniques = _nanunique(values)
+        u_mask = _get_mask(uniques, missing_value)
+        uniques = uniques[~u_mask]
+        check_values = False
+
+    if encode:
+        if check_values:
+            unique_v = _nanunique(values)
+            if is_scalar_nan(uniques[-1]) and is_scalar_nan(unique_v[-1]):
+                unique_v = unique_v[:-1]
+            unseen = np.setdiff1d(unique_v, uniques)
+            unseen = unseen[~_get_mask(unseen, missing_value)]
+            if unseen:
+                raise ValueError("y contains previously unseen labels: %s"
+                                 % str(unseen))
+
+        encoded = np.searchsorted(uniques, values)
+        e_mask = _get_mask(values, missing_value)
+        return uniques, encoded, e_mask
+    else:
+        return uniques
+
+
+def _nanencode_python(values, uniques=None, encode=False,
+                      missing_value=None):
+    if uniques is None:
+        uniques = set(values)
+        uniques.discard(missing_value)
+        # Put None at the end of the sort, similar to numpy.unique
+        uniques = sorted(uniques, key=lambda x: (x is None, x))
+        uniques = np.array(uniques, dtype=values.dtype)
+    if encode:
+        # Use index -1 so that the number encoding will not cause failures
+        # if used by the consumer for indexing. It will still fail when used
+        # for indexing an empty array but so is any other index.
+        nan_index = -1
+        table = {val: i for i, val in enumerate(uniques)}
+        table[missing_value] = nan_index
+        try:
+            encoded = np.array([table[v] for v in values])
+            encoded_nan = (encoded == nan_index)
```

Comment:
rm parentheses

Reply:
But I suspect the caller should do this instead of it being returned from here.

Reply:
The ```_nanencode_numpy``` returns a mask so it's to have a uniform interface between it and trhe ```nanencode_python```

Reply:
I suppose I'm okay with it, but I think we wouldn't need it if both `_nanencode_numpy` and `_nanencode_python` used -1 to represent missingness?

Reply:
Since I am using ```searchsorted``` for the numpy version, there is no way to easily make it as -1. I can get the mask and set it to -1 afterwards. However, I feel like the consumer of the function will probably have more use for the mask since they will use it to change how the nans are encoded.

Reply:
Yeah, I realise there's no way to easily make it -1 (or nan) other than setting the value given the mask. I really would need to see them being used in OneHotEncoder/OrdinalEncoder to understand whether this is beneficial.